### PR TITLE
Fix archival activities error handling

### PR DESCRIPTION
**Pull request by yycptt**
**What changed?**
- Do not mark all activities errors as non-retryable 😞 
- Change the condition for what errors is non-retryable for workflow deletion, more specifically, workflow not ready error and context errors (like context deadline exceeded, since each new attempt get a new startToClose timeout) should be retryable.

**Why?**
- Bug fix

**How did you test it?**
- Tested locally
- Updated unit test

**Potential risks**

**Is hotfix candidate?**
- Yes.

### CODE REVIEW

File: service/worker/archiver/activities.go [hunk 1]
```diff
@@ -123,9 +126,15 @@ func deleteHistoryActivity(ctx context.Context, request ArchiveRequest) (err err
 	if err == nil {
 		return nil
 	}
-	logger := tagLoggerWithHistoryRequest(tagLoggerWithActivityInfo(container.Logger, activity.GetInfo(ctx)), &request)
-	logger.Error("failed to delete history events", tag.Error(err))
-	if !common.IsPersistenceTransientError(err) {
+
+	if _, ok := err.(*serviceerror.WorkflowNotReady); !ok {
+		logger := tagLoggerWithHistoryRequest(tagLoggerWithActivityInfo(container.Logger, activity.GetInfo(ctx)), &request)
```
Comment (yycptt) [inline service/worker/archiver/activities.go:131]: Suppress logs for notReady error which is expected for deleteWorkflowExecution call. The error will still be logged if all retry fails.

**Review by yiminc — APPROVED** [review]

File: service/worker/archiver/activities.go [hunk 2]
```diff
@@ -123,9 +126,15 @@ func deleteHistoryActivity(ctx context.Context, request ArchiveRequest) (err err
...
+		logger.Error("failed to delete workflow execution", tag.Error(err))
+	}
+
+	if !common.IsServiceTransientError(err) &&
+		!common.IsContextDeadlineExceededErr(err) &&
+		!common.IsContextCanceledErr(err) {
```
Comment (yiminc) [inline service/worker/archiver/activities.go:137]: WorkflowNotReady will pass this check and return errDeleteNonRetryable?
Also IsServiceTransientError() already checked IsContextDeadlineExceededErr and IsContextCanceledErr.

File: service/worker/archiver/handler.go [hunk 3]
```diff
@@ -167,7 +167,7 @@ func (h *handler) handleHistoryRequest(ctx workflow.Context, request *ArchiveReq
 	localActCtx := workflow.WithLocalActivityOptions(ctx, lao)
 	err = workflow.ExecuteLocalActivity(localActCtx, deleteHistoryActivity, *request).Get(localActCtx, nil)
 	if err != nil {
-		logger.Error("deleting history failed, this means zombie histories are left", tag.Error(err))
+		logger.Error("deleting workflow execution failed all retires, skip workflow deletion", tag.Error(err))
```
Comment (yux0) [inline service/worker/archiver/handler.go:170]: Will the workflow data stay in Db if it just skip the deletion?

Comment (yycptt) [reply service/worker/archiver/handler.go:170]: Yes. Archival workflow will only retry up to 5min for both uploading and deletion and then give up. This is the limitation of the existing archival design. The issue will be gone once we have a separate archival queue.

For now, user should monitor the metrics for archival delete non-retryable error and use admin wf del command to manually delete those workflows from DB.

**Review by yiminc — APPROVED** [review]

File: service/worker/archiver/activities.go [hunk 4]
```diff
@@ -62,8 +63,9 @@ func uploadHistoryActivity(ctx context.Context, request ArchiveRequest) (err err
 		if err != nil {
 			if err.Error() == errUploadNonRetryable.Error() {
 				scope.IncCounter(metrics.ArchiverNonRetryableErrorCount)
+			} else {
+				err = temporal.NewApplicationError(err.Error(), "", nil)
```
Comment (alexshtin) [inline service/worker/archiver/activities.go:67]: This is default `error` to `ApplicationError` conversion. I would just return `err` here.
```suggestion
```

Comment (alexshtin) [reply service/worker/archiver/activities.go:67]: And everywhere bellow.

**Review by alexshtin — APPROVED** [review]

File: service/worker/archiver/activities.go (hunk 4 above)
Comment (yycptt) [reply service/worker/archiver/activities.go]: Makes sense. Then no change is needed to `err` and now the defer looks like the following:

```
	defer func() {
		sw.Stop()
		if err == errUploadNonRetryable {
			scope.IncCounter(metrics.ArchiverNonRetryableErrorCount)
		}
	}()
```

**Pull request merged**